### PR TITLE
Blog App: validations, model specs, and n+1 problems

### PR DESCRIPTION
**Pull request by KingsleyIbe**
### In this milestone, I implemented the following:

- Added validation
- Added unit specs
- Created controller actions

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,9 +1,38 @@
 class PostsController < ApplicationController
   def index
-    @posts = Post.where(author_id: params[:user_id])
+    @user = User.find(params[:user_id])
+    @posts = @user.posts.includes(:comments)
   end
 
   def show
     @post = Post.find(params[:id])
+    @user = @post.user
```
Comment (ahmetbozaci) [inline app/controllers/posts_controller.rb:9]: - [ ] Please fix the error. It happens when ı click the post on the user page.

image

File: spec/models/comment_spec.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      expect(Post.find(1).comments_counter).to_not eq(prev)
+    end
+
+    it 'check if it increases the comments' do
+      prev = Post.find(1).comments_counter
+      subject.update_comments_counter
+      expect(Post.find(1).comments_counter).to eq(prev + 1)
+    end
+  end
+end
```
Comment (ahmetbozaci) [inline spec/models/comment_spec.rb:25]: - [ ] You should fix the rspec errors.

image

File: spec/models/like_spec.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+    before { subject.save }
+
+    it 'check if it increases the posts' do
+      prev = Post.find(1).likes_counter
+      subject.update_likes_counter
+      expect(Post.find(1).likes_counter).to eq(prev + 1)
+    end
+  end
+end
```
Comment (ahmetbozaci) [inline spec/models/like_spec.rb:15]: - [ ] You should fix the rspec errors.

image

File: spec/models/post_spec.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    end
+
+    it 'check if it increases the posts' do
+      subject.user = User.find(1)
+      prev = User.find(1).posts_counter
+      subject.update_posts_counter
+      expect(User.find(1).posts_counter).to eq(prev + 1)
+    end
+  end
+end
```
Comment (ahmetbozaci) [inline spec/models/post_spec.rb:39]: - [ ] You should fix the rspec errors.

image

**Review by ahmetbozaci — CHANGES_REQUESTED** [review]
# STATUS: Changes Required  ♻️


 @KingsleyIbe  ✋🏼 

#### Your project is great. You are almost done. However, there are some issues you should fix. Please fix them then your project will be approved.


### ♻️ Request changes ♻️

- [ ] _Check the comments under the review._
- [ ] You should also add tests for the user model. [Requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec_n%2B1.md#:~:text=Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.)
- [ ] You should add `comments#create`, and `likes#create` actions use flash messages for successful and not successful path. That message is rendered in the view. Please check in the view it if flash message is work or not for post, like and comment create methods.

___

### Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
- [ ] I cant check if post create form flash message work or not because ı cant finds the post form in the page. Please add link for create form in the page.This form should be added to the `form milestone.` I make it as an optional but it should added before. Please add it.
- [ ] Also ı cant find comment form in the page. Please add it and check if flash message work or not.

___

### Cheers and Happy coding! 💻 👏
almost


___

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification._

_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._


___

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form.](https://airtable.com/shrQAqnBwek5a0O0s)_

[GitHub](https://github.com/ahmetbozaci) | [LinkedIn](https://linkedin.com/in/ahmetbozaci) | [Twitter](https://twitter.com/ahmtbozaci)

_Reviewed by [Ahmet Bozacı ](https://github.com/ahmetbozaci)_

File: spec/models/post_spec.rb (hunk 4 above)
Comment (KingsleyIbe) [reply spec/models/post_spec.rb]: Hello @ahmetbozaci 

Thank you for finding time to review my project.
All requested changes have been made.
WhatsApp Image 2022-04-02 at 5 22 49 PM

**Review by descholar-ceo — APPROVED** [review]
Hi @KingsleyIbe , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

- [ ] I would recommend you to integrate rspec testing with github actions, so that it becomes easy to see test output right on your PRs

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by KingsleyIbe** [comment]
Hello @descholar-ceo 

Thanks for reviewing and approving my project.
Also, thanks for your recommendation, I would checkout how to achieve that.

**Pull request merged**